tensorboard/notebook.py

```diff
@@ -0,0 +1,336 @@
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import datetime
+import shlex
+import sys
+
+from tensorboard import manager
+
+
+# See `_get_context` for the meanings of these values.
+_CONTEXT_COLAB = "_CONTEXT_COLAB"
+_CONTEXT_IPYTHON = "_CONTEXT_IPYTHON"
+_CONTEXT_NONE = "_CONTEXT_NONE"
+
+
+def _get_context():
+  """Determine the most specific context that we're in.
+
+  Returns:
+    _CONTEXT_COLAB: If in Colab with an IPython notebook context.
+    _CONTEXT_IPYTHON: If not in Colab, but we are in an IPython notebook
+      context (e.g., from running `jupyter notebook` at the command
+      line).
+    _CONTEXT_NONE: Otherwise (e.g., by running a Python script at the
+      command-line or using the `ipython` interactive shell).
+  """
+  # In Colab, the `google.colab` module is available, but the shell
+  # returned by `IPython.get_ipython` does not have a `get_trait`
+  # method.
+  try:
+    import google.colab
+    import IPython
+  except ImportError:
+    pass
+  else:
+    if IPython.get_ipython() is not None:
+      # We'll assume that we're in a notebook context.
+      return _CONTEXT_COLAB
+
+  # In an IPython command line shell or Jupyter notebook, we can
+  # directly query whether we're in a notebook context.
+  try:
+    import IPython
+  except ImportError:
+    pass
+  else:
+    ipython = IPython.get_ipython()
+    if ipython is not None and ipython.has_trait("kernel"):
+      return _CONTEXT_IPYTHON
+
+  # Otherwise, we're not in a known notebook context.
+  return _CONTEXT_NONE
+
+
+def load_ipython_extension(ipython):
+  """IPython API entry point.
+
+  See:
+    https://ipython.readthedocs.io/en/stable/config/extensions/index.html
+  """
+  _register_magics(ipython)
+
+
+def _register_magics(ipython):
+  """Register IPython line/cell magics.
+
+  Args:
+    ipython: An `InteractiveShell` instance.
+  """
+  ipython.register_magic_function(
+      _start_magic,
+      magic_kind="line",
+      magic_name="tensorboard",
+  )
+
+
+def _start_magic(line):
+  """Implementation of the `%tensorboard` line magic."""
+  return start(line)
+
+
+def start(args_string):
+  """Launch and display a TensorBoard instance as if at the command line.
+
+  Args:
+    args_string: Command-line arguments to TensorBoard, to be
+      interpreted by `shlex.split`: e.g., "--logdir ./logs --port 0".
+      Shell metacharacters are not supported: e.g., "--logdir 2>&1" will
+      point the logdir at the literal directory named "2>&1".
+  """
+  context = _get_context()
+  try:
+    import IPython
+    import IPython.display
+  except ImportError:
+    IPython = None
+
+  if context == _CONTEXT_NONE:
+    handle = None
+    print("Launching TensorBoard...")
+  else:
+    handle = IPython.display.display(
+        IPython.display.Pretty("Launching TensorBoard..."),
+        display_id=True,
+    )
+  def print_or_update(message):
```

Comment:
nit: I suggest adding a newline above this for readability.

Reply:
Sure.